### ISSUE TITLE
New Rule Proposal: Consistent async function naming

### ISSUE THREAD

**Issue by dtjohnson** [eslint/eslint#8531]
**Please describe what the rule should do:**

This rule enforces that `async` functions are named with an Async suffix and also that any function named with an Async suffix be marked as `async`. It is a common practice for functions that return promises to be named with an Async suffix to indicate to the caller that they should handle a promise. (They either need to `await` the result, call `.then`, etc.) In fact, this is also the [standard naming convention in .NET](https://msdn.microsoft.com/en-us/library/hh873175.aspx), which has a similar async/await pattern. 

**What category of rule is this? (place an "X" next to just one item)**

[x] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Incorrect versions:
```js
async function doSomething() {
    // ...
}

function doSomethingElseAsync() {
    // ...
}
```

Correct versions:
```js
async function doSomethingAsync() {
    // ...
}

async function doSomethingElseAsync() {
    // ...
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

The rule inforces consistency of a very common practice (from even before the `async` keyword existed). It is general, library agnostic, and shouldn't conflict with other rules.

**Comment by not-an-aardvark** [comment]
Thanks for the rule suggestion.

Personally, I'm not sure about this -- I think this seems like this is an opinionated preference and would be a better fit for a plugin. However, let's see what the rest of the team thinks.

I think this rule would also be possible to simulate with a [`no-restricted-syntax`](http://eslint.org/docs/rules/no-restricted-syntax) configuration.

**Comment by dtjohnson** [comment]
The naming convention could also be configurable. The "Async" suffix is just what has been popularized by [.NET](https://msdn.microsoft.com/en-us/library/hh873175.aspx) and [Bluebird's promisifyAll](http://bluebirdjs.com/docs/api/promise.promisifyall.html). The key concept is some naming convention that makes it clear the return is async. This convention would help prevent the very common mistake of not handling a returned promise. Consider this example with a sync and async function:
```js
function doEverything() {
    doSomethingAsync();
    doSomethingElse();
}
```
It's clear from the name `doSomethingAsync` that this code likely has a bug. The code should be:
```js
async function doEverythingAsync() {
    await doSomethingAsync();
    doSomethingElse();
}
```

As a separate but related issue, it would be nice to have a supplemental rule that would enforce that function calls ending in "Async" should be `await`ed, `.then`ed, or commented that they are purposely not being handled. That could be tricky though as there are lots of other legitimate things you could do with a promise like `Promise.all`, but it could prevent a lot of bugs.

**Comment by platinumazure** [comment]
I like @not-an-aardvark's idea of using `no-restricted-syntax` with selectors to achieve this.

Maybe something like:

```json
{
    "no-restricted-syntax": [
        "error",
        {
            "selector": "FunctionDeclaration[async=false][id.name=/Async$/]",
            "message": "Function ending in 'Async' must be declared async"
        },
        {
            "selector": "FunctionDeclaration[async=true][id.name!=/Async$/]",
            "message": "Async function declaration name must end in 'Async'"
        }
    ]
}
```

(ESQuery has tests which use a `!=` operator and a regex, so hopefully this will work.)

**Comment by ilyavolodin** [comment]
I agree with @not-an-aardvark I think this is pretty opinionated rule for the core, and should be part of a plugin (or as @platinumazure showed, should just be part of the configuration). Although I'm biased. I hate prefixes and postfixes for variable/function names.

**Comment by dtjohnson** [comment]
Seems fair that you don't want the core to be so opinionated. @not-an-aardvark's suggestion and @platinumazure's example were a great start. However, there are a number of cases to cover:
```js
async function fooAsync() {
}

function foo() {
}

const barAsync = async function () {
};

const bar = function () {
};

const bazAsync = async () => {
};

const baz = () => {
};

class Foo {
    async fooAsync() {
    }

    foo() {
    }
}

const obj = {
    async fooAsync() {
    },

    foo() {
    },

    barAsync: async function () {
    },

    bar: function () {
    },

    bazAsync: async () => {
    },

    baz: () => {
    }
};
```

I expanded the rule a bit to cover all of these cases and it seems to catch all of these:
```js
{
  "no-restricted-syntax": [
    "error",
    {
      "selector": "FunctionDeclaration[async=false][id.name=/Async$/]",
      "message": "Function ending in 'Async' must be declared async"
    },
    {
      "selector": "FunctionDeclaration[async=true][id.name!=/Async$/]",
      "message": "Async function name must end in 'Async'"
    },
    {
      "selector": "MethodDefinition[value.async=false][key.name=/Async$/]",
      "message": "Method ending in 'Async' must be declared async"
    },
    {
      "selector": "MethodDefinition[value.async=true][key.name!=/Async$/]",
      "message": "Async method name must end in 'Async'"
    },
    {
      "selector": "Property[value.type=/FunctionExpression$/][value.async=false][key.name=/Async$/]",
      "message": "Function ending in 'Async' must be declared async"
    },
    {
      "selector": "Property[value.type=/FunctionExpression$/][value.async=true][key.name!=/Async$/]",
      "message": "Async function name must end in 'Async'"
    },
    {
      "selector": "VariableDeclarator[init.type=/FunctionExpression$/][init.async=false][id.name=/Async$/]",
      "message": "Function ending in 'Async' must be declared async"
    },
    {
      "selector": "VariableDeclarator[init.type=/FunctionExpression$/][init.async=true][id.name!=/Async$/]",
      "message": "Async function name must end in 'Async'"
    }
  ],
}
```

If you don't feel something like this should be in the core, then at least this issue serves as an example for other folks who want to follow this async pattern. Thanks!

**Comment by platinumazure** [comment]
@dtjohnson Nice work! You might want to add selectors for AssignmentExpression (just in case) for situations like the below:

```js
let funcAsync;

funcAsync = async function() { };
```

**Comment by not-an-aardvark** [comment]
Closing because it's possible to do this with selector, and there doesn't seem to be any interest from the ESLint team in adding this to core.
